torch/autograd/gradcheck.py

```diff
@@ -806,41 +886,88 @@ def wrapped_fn(*inputs):
         return wrapped_fn
     return apply_to_c_outs(fn, torch.real), apply_to_c_outs(fn, torch.imag)
 
+def _real_and_imag_input(fn):
+    # returns new functions that take real inputs instead of complex inputs and compute fn(x + 0 * 1j)
+    # and f(x * 1j).
+    def apply_to_c_inps(fn, fn_to_apply):
+        def wrapped_fn(*inputs):
+            new_inputs = (fn_to_apply(inp) if inp.is_complex() else inp for inp in inputs)
+            return _as_tuple(fn(*new_inputs))
+        return wrapped_fn
+    return apply_to_c_inps(fn, lambda x: x + 0 * 1j), apply_to_c_inps(fn, lambda x: x * 1j)
+
 
 def _gradcheck_real_imag(gradcheck_fn, func, func_out, tupled_inputs, outputs, eps, rtol,
-                         atol, check_grad_dtypes, nondet_tol, complex_indices, any_outputs_complex):
-    if any_outputs_complex:
-        real_fn, imag_fn = _real_and_imag(func, outputs)
+                         atol, check_grad_dtypes, check_forward_ad, nondet_tol):
```

Comment:
Feels like there's some logic repeated here between the two blocks, but not entirely sure there's an easy way to fix it.

Reply:
I don't think so, because we now wrap the input and not the output and we need to pass slightly different flags.
I feel like abstracting that would just create more boilerplate code.